### PR TITLE
🌱 enable godot on top-level comments

### PR DESCRIPTION
**Pull request by sbueringer**
**What this PR does / why we need it**:

Goal of the PR in its current state is to discuss if we want to enable godot also on toplevel comments (default is declarations).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

File: bootstrap/kubeadm/internal/cloudinit/cloudinit.go [hunk 1]
```diff
@@ -28,7 +28,7 @@ import (
 
 const (
 	standardJoinCommand = "kubeadm join --config /run/kubeadm/kubeadm-join-config.yaml %s"
-	// sentinelFileCommand writes a file to /run/cluster-api to signal successful Kubernetes bootstrapping in a way that
+	// SentinelFileCommand writes a file to /run/cluster-api to signal successful Kubernetes bootstrapping in a way that
```
Comment (sbueringer) [inline bootstrap/kubeadm/internal/cloudinit/cloudinit.go:31]: godot doesn't detect that it's a variable name

File: cmd/clusterctl/client/cluster/upgrader.go [hunk 2]
```diff
@@ -187,7 +187,7 @@ func (u *providerUpgrader) ApplyCustomPlan(upgradeItems ...UpgradeItem) error {
 }
 
 // getUpgradePlan returns the upgrade plan for a specific set of providers/contract
-// NB. this function is used both for upgrade plan and upgrade apply.
+// NB. This function is used both for upgrade plan and upgrade apply.
```
Comment (sbueringer) [inline cmd/clusterctl/client/cluster/upgrader.go:190]: godot enforces capital letters after things like `NB.` or `E.g.`

File: cmd/clusterctl/client/repository/client.go [hunk 3]
```diff
@@ -131,7 +131,7 @@ func newRepositoryClient(provider config.Provider, configClient config.Client, o
 }
 
 // Repository defines the behavior of a repository implementation.
-// clusterctl is designed to support different repository types; each repository implementation should be aware of
+// Clusterctl is designed to support different repository types; each repository implementation should be aware of
```
Comment (sbueringer) [inline cmd/clusterctl/client/repository/client.go:134]: godot obviously doesn't know that it's the name of a binary

File: hack/boilerplate/test/fail.go [hunk 4]
```diff
@@ -1,7 +1,7 @@
 /*
 Copyright 2014 The Kubernetes Authors.
 
-fail
+Fail
```
Comment (sbueringer) [inline hack/boilerplate/test/fail.go:4]: That would be something we have to exclude if we want to keep the capital letter linting

**Review by sbueringer — COMMENTED** [review]
I think capitalization linting doesn't make sense as it has too many false positives.

The actual `dot` part looks fine to me. 

I would enable toplevel linting but not capitalization
```yaml
    #   declarations - for top level declaration comments (default);
    #   toplevel     - for top level comments;
    #   all          - for all comments.
    scope: toplevel
    exclude:
    - '^ \+.*'
    - '^ ANCHOR.*'
```

WDYT?

**Comment by fabriziopandini** [comment]
This goes in the direction of more consistency across the code base, so I'm +1

File: hack/boilerplate/test/fail.go (hunk 4 above)
Comment (stmcginnis) [reply hack/boilerplate/test/fail.go:4]: Yeah, might be good to just add all `../test/*` paths to the ignore list.

**Review by stmcginnis — COMMENTED** [review]
I agree the punctuation is more useful than the capitalization.

It's probably OK to enable "all", but the only concern I would have with that is it's not obvious to new developers coming in that the linting will flag comments starting with the name of a privately scoped variable (lowercase) as an issue, so it could cause some frustration when linting checks fail. Not a huge deal, but something to consider.

File: controlplane/kubeadm/controllers/consts.go [hunk 5]
```diff
@@ -19,15 +19,15 @@ package controllers
 import "time"
 
 const (
-	// deleteRequeueAfter is how long to wait before checking again to see if
+	// DeleteRequeueAfter is how long to wait before checking again to see if
```
Comment (vincepri) [inline controlplane/kubeadm/controllers/consts.go:22]: These shouldn't be capitalized

File: .golangci.yml [hunk 6]
```diff
@@ -39,6 +39,15 @@ linters:
...
 linters-settings:
+  godot:
+    #   declarations - for top level declaration comments (default);
+    #   toplevel     - for top level comments;
+    #   all          - for all comments.
+    scope: toplevel
+    exclude:
+    - '^ \+.*'
+    - '^ ANCHOR.*'
+    capital: true
```
Comment (vincepri) [inline .golangci.yml:50]: Do we need this? It seems flaky imo and lots of friction. The benefit of having a capital letter after a dot is small compared to a smoother new contributor experience

Comment (sbueringer) [reply .golangci.yml:50]: I would drop capitalization as stated in: https://github.com/kubernetes-sigs/cluster-api/pull/5591#pullrequestreview-798648322

Just wanted to wait for additional opinions.

File: hack/boilerplate/test/fail.go (hunk 4 above)
Comment (sbueringer) [reply hack/boilerplate/test/fail.go:4]: Should be fine now, I dropped the capitalization.

**Comment by sbueringer** [comment]
Dropped capitalization.

File: .golangci.yml (hunk 6 above)
Comment (sbueringer) [reply .golangci.yml]: Done

**Review by vincepri — APPROVED** [review]
/approve

**Pull request merged**